src/enumerate.rs

```diff
@@ -1 +1,202 @@
 //! State enumeration
+use crate::dfa;
+use crate::dfa::Dfa;
+use crate::dfa::State;
+use crate::eval;
+use crate::eval::RuntimeError;
+use crate::eval::Value;
+use crate::syntax::*;
+use std::collections::BTreeMap;
+
+type Env = BTreeMap<Id, Value>;
+
+/**
+ *
+ */
+pub fn enumerate(program: &Program, _input: &str) -> Result<Dfa<Symbol>, RuntimeError> {
+    // keep track of visited states and their environments
+    // let mut state_lookup: Map<State, Env> = Map::new();
+    let mut state_lookup: Map<State, Env> = Map::new();
+    // the transitions between states
+    let mut trans: Map<State, Map<Symbol, State>> = Map::new();
+    //state names
+    let mut names: Map<State, String> = Map::new();
+
+    let mut env_lookup: BTreeMap<Env, State> = BTreeMap::new();
+
+    //get initial state
+    let mut init_e = eval::init_env(program);
+    for stmt in &program.start {
+        eval::eval_stmt(stmt, &mut init_e, &program)?;
+    }
+
+    //add inital state
+    let init_s = dfa::State::fresh();
+    state_lookup.insert(init_s, init_e.clone());
+    env_lookup.insert(init_e.clone(), init_s);
+
+    let mut workqueue: Vec<State> = Vec::new();
+    workqueue.insert(0, init_s);
+    let mut accepting: Set<State> = Set::new();
+    // check acceptance for init env
+
+    while workqueue.is_empty() == false {
+        let s = workqueue.pop().unwrap();
+        let mut s_edges: Map<Symbol, State> = Map::new();
+        let curr_env = state_lookup.get(&s).unwrap().clone();
+        for sym in &program.alphabet {
+            let mut env_clone = curr_env.clone();
+            eval::eval_action(program, &mut env_clone, sym)?;
+
+            if let Some(_) = &program.action.0 {
+                env_clone.remove(&program.action.0.unwrap());
+            }
+
+            let mut new = true;
+            if env_lookup.contains_key(&env_clone) {
+                new = false;
+            }
+
+            let t = env_lookup
+                .entry(env_clone.clone())
+                .or_insert_with(|| dfa::State::fresh());
+
+            s_edges.insert(*sym, t.clone());
+            if new == true {
+                state_lookup.insert(*t, env_clone.clone());
+                workqueue.insert(workqueue.len(), t.clone());
+            }
+        }
+        trans.insert(s, s_edges);
+    }
+    for st in state_lookup.keys().clone() {
+        let accept = eval::eval_expr(
+            &program.accept.clone(),
+            &state_lookup.get(&st).unwrap().clone(),
+            &program,
+        )?;
+        if accept == Value::Bool(true) {
+            // assuming that all the accept statments of programs are bools
+            accepting.insert(*st);
+        }
+
+        let state_env = state_lookup.get(st).unwrap();
+        let f = state_env.first_key_value().unwrap();
+        names.insert(*st, format!("{} = {}", f.0, f.1).to_string());
```

Comment:
this is bad state naming: what if there are multiple variables in the program? you should create a proper name, something like `state_env.iter().map(|(x, v)| format!("{x} = {v}")).join(", ")`.